clang/lib/CodeGen/CGHLSLRuntime.h

```diff
@@ -30,22 +30,36 @@
 #include <optional>
 #include <vector>
 
+#define GENERATE_HLSL_INTRINSIC_FUNCTION_DEFAULT(FunctionName,                 \
+                                                 IntrinsicPostfix)             \
+  GENERATE_HLSL_INTRINSIC_FUNCTION(FunctionName, IntrinsicPostfix, 1, 1)
+
 // A function generator macro for picking the right intrinsic
 // for the target backend
-#define GENERATE_HLSL_INTRINSIC_FUNCTION(FunctionName, IntrinsicPostfix)       \
+#define GENERATE_HLSL_INTRINSIC_FUNCTION(FunctionName, IntrinsicPostfix,       \
+                                         IncludeDXIL, IncludeSPIRV)            \
   llvm::Intrinsic::ID get##FunctionName##Intrinsic() {                         \
     llvm::Triple::ArchType Arch = getArch();                                   \
     switch (Arch) {                                                            \
-    case llvm::Triple::dxil:                                                   \
-      return llvm::Intrinsic::dx_##IntrinsicPostfix;                           \
-    case llvm::Triple::spirv:                                                  \
-      return llvm::Intrinsic::spv_##IntrinsicPostfix;                          \
+      /* Include DXIL case only if IncludeDXIL is true */                      \
+      IF_INCLUDE(IncludeDXIL, case llvm::Triple::dxil                          \
+                 : return llvm::Intrinsic::dx_##IntrinsicPostfix;)             \
+      /* Include SPIRV case only if IncludeSPIRV is true */                    \
+      IF_INCLUDE(IncludeSPIRV, case llvm::Triple::spirv                        \
+                 : return llvm::Intrinsic::spv_##IntrinsicPostfix;)            \
+                                                                               \
     default:                                                                   \
       llvm_unreachable("Intrinsic " #IntrinsicPostfix                          \
                        " not supported by target architecture");               \
     }                                                                          \
   }
 
+#define IF_INCLUDE(Condition, Code) IF_INCLUDE_IMPL(Condition, Code)
+#define IF_INCLUDE_IMPL(Condition, Code) IF_INCLUDE_##Condition(Code)
+
+#define IF_INCLUDE_1(Code) Code
+#define IF_INCLUDE_0(Code)
```

Comment:
I feel like this is a bit of extra preprocessor complexity. Instead of having `GENERATE_HLSL_INTRINSIC_FUNCTION` change to take arguments, we could just define target-specific macros (i.e. `GENERATE_HLSL_DX_INTRINSIC_FUNCTION` and `GENERATE_HLSL_SPV_INTRINSIC_FUNCTION`).

I also generally think putting `DEFAULT` in the name of something is a bit of a red flag since you're making the name longer, but not adding relevant context.

Reply:
that would mean a new macro per backend. Also the point of the macro was to make one function that could toggle between all the different target intrinsics our language supports. If we have one generate per DX or SPV that's no better than directly passing the target intrinsic id.

Reply:
I don't see how creating target specific macros is better than just using `llvm::Intrinsic::spv_*` or `llvm::Intrinsic::dx_*`. The idea behind making it arguments is that if we have more targets after spv and dx we aren't adding an  `N` number of new macros its just one more argument and one more `IF_INCLUDE` case block. 

instead of `DEFAULT` I could rename to `ALL`? 

- I could also make `GENERATE_HLSL_DX_INTRINSIC_FUNCTION`  --> `GENERATE_HLSL_INTRINSIC_FUNCTION(FunctionName, IntrinsicPostfix, 1, 0)` 
- and `GENERATE_HLSL_SPV_INTRINSIC_FUNCTION` --> `GENERATE_HLSL_INTRINSIC_FUNCTION(FunctionName, IntrinsicPostfix, 0, 1)`

Reply:
I believe the C preprocessor is the wrong tool to do what you're trying to do here. I think this makes the code significantly harder to read, understand and debug.

There are other approaches we could take like constexpr template abstractions which would be easier to read, write and maintain than nested C preprocessor macros with dependent expansions.

Reply:
Fair enough closing this PR.